kale/task.py

```diff
@@ -177,28 +195,62 @@ def run(self, *args, **kwargs):
         4) call super() at the start of _post_run override.
         5) Subclass's override _post_run logic.
         """
+        if not settings.is_multiprocess_mode():
+            self._run_single(*args, **kwargs)
+        else:
+            self._run_multi(*args, **kwargs)
 
-        self._setup_task_environment()
-        self._pre_run(*args, **kwargs)
+    def _run_single(self, *args, **kwargs):
+        """Runs the task in single process mode.
 
+        This will use SIGALRM to handle timeouts.
+
+        This calls _task_wrapper, which handles all task logic.
+        """
+        self._record_start()
         try:
-            # This raises an exception if the task should not be attempted.
-            # This enables tasks to be blacklisted by ID or type.
-            self._check_blacklist(*args, **kwargs)
-            self.run_task(*args, **kwargs)
-        except Exception as exc:
-            # Record latency here.
-            self._end_time = time.time()
-            self._task_latency_sec = self._end_time - self._start_time
-
-            # Cleanup the environment this task was running in right away.
-            self._clean_task_environment(task_id=self.task_id,
-                                         task_name=self.task_name, exc=exc)
-            raise
-
-        self._post_run(*args, **kwargs)
-        self._clean_task_environment(task_id=self.task_id,
-                                     task_name=self.task_name)
+            def _handle_timeout(signum, frame):
+                raise exceptions.TimeoutException(os.strerror(errno.ETIME))
+
+            # Use sigalrm to handle timeout.
+            signal.signal(signal.SIGALRM, _handle_timeout)
+            signal.alarm(self.time_limit)
+            exc = _task_wrapper(self, args, kwargs)
+            if exc is not None:
+                raise exc
+        finally:
+            self._record_end()
+            # Remove handler.
+            signal.signal(signal.SIGALRM, signal.SIG_DFL)
+            # Remove signal.
+            signal.alarm(0)
+
+    def _run_multi(self, *args, **kwargs):
+        """Runs the task in a subprocess.
+
+        The subprocess runs _task_wrapper, which handles all task logic.
+
+        Because the start and end time must be known to the worker,
+        _record_start() and _record_end() are run in the context of the parent process.
+        """
+        self._record_start()
+        try:
+            process = Task._task_process_pool.apply_async(_task_wrapper, args=(self, args, kwargs))
+            exc = process.get(self.time_limit)
+            if exc is not None:
+                exc.reraise()
+        except multiprocessing.TimeoutError:
+            # Hard-terminate the subprocess.
+            Task._task_process_pool.terminate()
+            Task._task_process_pool = multiprocessing.Pool(1)
```

Comment:
As mentioned above, move the initialization to the beginning of this function so we don't need to remember to reinitialize every time we terminate.